Add method to remove component and all required components for removed component

rewin123 opened the pull request:
## Objective
The new Required Components feature (#14791) in Bevy allows spawning a fixed set of components with a single method with cool require macro. However, there's currently no corresponding method to remove all those components together. This makes it challenging to keep insertion and removal code in sync, especially for simple using cases.
```rust
#[derive(Component)]
#[require(Y)]
struct X;

#[derive(Component, Default)]
struct Y;

world.entity_mut(e).insert(X); // Spawns both X and Y
world.entity_mut(e).remove::<X>(); 
world.entity_mut(e).remove::<Y>(); // We need to manually remove dependencies without any sync with the `require` macro
```
## Solution
Simplifies component management by providing operations for removal required components.
This PR introduces simple 'footgun' methods to removes all components of this bundle and its required components.

Two new methods are introduced:
For Commands:
```rust
commands.entity(e).remove_with_requires::<B>();
```
For World:
```rust
world.entity_mut(e).remove_with_requires::<B>();
```

For performance I created new field in Bundels struct. This new field "contributed_bundle_ids" contains cached ids for dynamic bundles constructed from bundle_info.cintributed_components()

## Testing
The PR includes three test cases:

1. Removing a single component with requirements using World.
2. Removing a bundle with requirements using World.
3. Removing a single component with requirements using Commands.
4. Removing a single component with **runtime** requirements using Commands

These tests ensure the feature works as expected across different scenarios.

## Showcase
Example:
```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
#[require(Y)]
struct X;

#[derive(Component, Default)]
#[require(Z)]
struct Y;

#[derive(Component, Default)]
struct Z;

#[derive(Component)]
struct W;

let mut world = World::new();

// Spawn an entity with X, Y, Z, and W components
let entity = world.spawn((X, W)).id();

assert!(world.entity(entity).contains::<X>());
assert!(world.entity(entity).contains::<Y>());
assert!(world.entity(entity).contains::<Z>());
assert!(world.entity(entity).contains::<W>());

// Remove X and required components Y, Z
world.entity_mut(entity).remove_with_requires::<X>();

assert!(!world.entity(entity).contains::<X>());
assert!(!world.entity(entity).contains::<Y>());
assert!(!world.entity(entity).contains::<Z>());

assert!(world.entity(entity).contains::<W>());
```

## Motivation for PR
#15580 

## Performance

I made simple benchmark
```rust
let mut world = World::default();
let entity = world.spawn_empty().id();

let steps = 100_000_000;

let start = std::time::Instant::now();
for _ in 0..steps {
    world.entity_mut(entity).insert(X);
    world.entity_mut(entity).remove::<(X, Y, Z, W)>();
}
let end = std::time::Instant::now();
println!("normal remove: {:?} ", (end - start).as_secs_f32());
println!("one remove: {:?} micros", (end - start).as_secs_f64() / steps as f64 * 1_000_000.0);

let start = std::time::Instant::now();
for _ in 0..steps {
    world.entity_mut(entity).insert(X);
    world.entity_mut(entity).remove_with_requires::<X>();
}
let end = std::time::Instant::now();
println!("remove_with_requires: {:?} ", (end - start).as_secs_f32());
println!("one remove_with_requires: {:?} micros", (end - start).as_secs_f64() / steps as f64 * 1_000_000.0);
```

Output:

CPU: Amd Ryzen 7 2700x

```bash
normal remove: 17.36135 
one remove: 0.17361348299999999 micros
remove_with_requires: 17.534006 
one remove_with_requires: 0.17534005400000002 micros
```

NOTE: I didn't find any tests or mechanism in the repository to update BundleInfo after creating new runtime requirements with an existing BundleInfo. So this PR also does not contain such logic.

## Future work (outside this PR)

Create cache system for fast removing components in "safe" mode, where "safe" mode is remove only required components that will be no longer required after removing root component.

SkiFire13 reviewed (commented):
> Simplifies component management by providing symmetrical operations for removal required components.

These are not entirely symmetrical, since if `X` requires `Y` and `Y` is already present then inserting `X` will not replace `Y`, but later removing `X` will remove `Y` as well.

In other words `.insert(X).remove_with_required::<X>()` is not a guaranteed noop.

I think this is fine, since there's probably no way to avoid this behaviour, but I think it should be mentioned in the documentation for `remove_with_required` as a potential pitfall.

tim-blackbird commented:
I'm concerned this might be too foot-gunny in the case of X and Y both requiring Z and then X is removed, leaving Y without Z.
It would be great if we could keep the required components if they were still required by something else.

alice-i-cecile commented:
I agree with @tim-blackbird; there's a serious risk that this ends up removing e.g. `Transform` by accident.

rewin123 commented:
Thank you all. I have read the comments and agree with all of them. I will then try to re-create the method to take into account the dependency tree of components on each other. And improve the documentation.

rewin123 commented:
@alice-i-cecile @tim-blackbird @SkiFire13 

I created requirement safe version of remove_with_requirement, updated PR description and improved docs. I would appreciate it if you could look at the PR again.

rewin123 commented:
> @alice-i-cecile @tim-blackbird @SkiFire13
> 
> I created component require safe version of remove_with_required function, updated PR description and improved docs. I would appreciate it if you could look at the PR again.

Also I added descent remove mode, which remove in recursive manner (and safe mode) all required components, bundle components and all components, which require bundle components.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
 
+    /// Removes all components of this [`Bundle`] and its required components that are not required by other components of this entity.
+    ///
+    /// This function can be noticeably slower than simple remove or retain functions because it dynamically determines which components
+    /// are still required by entity components outside of the [`Bundle`].
```
Given this observation, would it make sense to cache this transition in the archetype graph?

rewin123 replied:
I probably should do some time measurements.... For a small number of components, the removal time will not be meaningfully increased, but caching will significant increase memory consumption

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
+        self
+    }
+
+    /// Removes the components of this [`Bundle`], components that require them, and their required components,
+    /// unless they are required by components outside of this removal set.
```
It took me a while (and the example) to understand what this function does, and even then I'm not sure what happens in all cases. It also seems to me this is going a bit out of scope. Is there an actual usecase for this?

rewin123 replied:
> It took me a while (and the example) to understand what this function does, and even then I'm not sure what happens in all cases. It also seems to me this is going a bit out of scope. Is there an actual usecase for this?

I can think of two examples of usage:
1. Complex status effects in games: When a character is affected by a status (e.g., poisoning), it may involve multiple related components. Different types of poisoning could require different visual effects, all depending on a core "Poisoned" component. When curing the poison, you want to remove not just the core component, but all related effect components. This function allows you to remove the entire "tree" of components related to the poison effect without having to manually track and remove each one.
2. Library cleanup for modular systems: In a scenario where you're developing a library (e.g., for animations), end users might extend your core components with their own custom components. When an animation completes and you want to clean up, using require_descendant_remove on the core animation component would also remove all user-defined components that depend on it, ensuring a thorough cleanup without leaving orphaned components behind.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
+
+        // SAFETY: the `BundleInfo` is initialized above
+        // We make clone of components array to not lock immutable access to world
+        let contributed_components = unsafe {
+            self.world
+                .bundles
+                .get_unchecked(bundle)
+                .contributed_components()
+                .to_vec()
+        };
+
+        let old_location = self.location;
+        let old_archetype = &mut self.world.archetypes[old_location.archetype_id];
+
+        let remaining_components = old_archetype
+            .components()
+            .filter(|c| !contributed_components.contains(c))
+            .collect::<Vec<_>>();
+
+        let remaining_bundle = self
+            .world
+            .bundles
+            .init_dynamic_info(components, &remaining_components);
+
+        // SAFETY: the `BundleInfo` is initialized above
+        let remaining_bundle_info = unsafe { self.world.bundles.get_unchecked(remaining_bundle) };
+
+        // Remove components required by remaining components from contributed_components of bundle T
+        let to_delete = contributed_components
+            .iter()
+            .filter(|c| !remaining_bundle_info.contributed_components().contains(c))
+            .copied()
+            .collect::<Vec<_>>();
```
I feel like the logic is becoming complex enough that it might need a bit more comments. From what I see the process is essentially this:

1. gather all the components that the `T` bundle would contribute (`T`'s components, their required components, required components of those required components, and so on...) in `contributed_components`;
    - these might however contains components that are required by other existing components, so they can't be removed immediately.
2. gather all the remaining components in `remaining_components`, that is all the components in this archetypes that are not included in 1.;
3. find all the components that these components would contribute with `remaining_bundle_info.contributed_components()`, these are those that we want to keep;
4. thus we have to remove all the others, which are computed in `to_delete`.

Might be worth comment where each of these steps happens.

Something similar, although even more complex, seems to be happening in `require_descendant_remove`, so it would be nice to add some comments there as well.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
+    /// This method is useful for removing a component along with its entire requirement tree and any
+    /// components that depend on them, while preserving components that are still required by parts of
+    /// the entity outside the removal tree.
+    pub fn require_descendant_remove<T: Bundle>(&mut self) -> &mut Self {
```
I'm not particularly fond of this name, as it clashes with the concept of descendant in `bevy_hierarchy`.

rewin123 replied:
Yes. I just checked and indeed descendal only destroys descendants without destroying the entity itself. I renamed the function that cleans up and up the require tree to avoid confusion

alice-i-cecile commented:
I think that this approach is better than the previous one, although I'm a bit nervous about adding a feature like this without user testing demonstrating that it's a problem.

rewin123 commented:
> I think that this approach is better than the previous one, although I'm a bit nervous about adding a feature like this without user testing demonstrating that it's a problem.
> 

Yes, I understand the concern. If this PR first needs to be validated by users, then I could put the PR implementation in a standalone crate, publish the crate after the release, and if there are any significant number of users of the crate, then reopen the PR afterwards.  Would that be a good solution?

rewin123 commented:
Based on discord discussion moving PR into footgun solution. Safe solution will be made in future
https://discord.com/channels/691052431525675048/1264881140007702558/1290788223861264447

rewin123 commented:
I want optimize remove_wire_requires. So move PR into draft

alice-i-cecile reviewed (commented):
Can you try using this PR to clean up the mess left behind in https://github.com/bevyengine/bevy/pull/15572#discussion_r1783376860 ? :)

rewin123 commented:
> Can you try using this PR to clean up the mess left behind in [#15572 (comment)](https://github.com/bevyengine/bevy/pull/15572#discussion_r1783376860) ? :)

Sure! I can do it in another PR after accepting this one, right?)

cart commented on crates/bevy_ecs/src/bundle.rs:
```diff
@@ -1346,11 +1349,39 @@ impl Bundles {
+            id
+        } else {
+            let explicit_bundle_id = self.register_info::<T>(components, storages);
+            // SAFETY: `explicit_bundle_id` is valid and defined above
+            let contributed_components = unsafe {
+                self.get_unchecked(explicit_bundle_id)
+                    .contributed_components()
+                    .to_vec()
+            };
+            let id = self.init_dynamic_info(components, &contributed_components);
```
This `to_vec` exists "just" for lifetime reasons. It is sound to pass the original vec into `init_dynamic_info`. Bundle init is performance critical enough to justify using some extra unsafe code here:
```suggestion
            let id = unsafe {
                let (ptr, len) = {
                    // SAFETY: `explicit_bundle_id` is valid and defined above
                    let contributed = self
                        .get_unchecked(explicit_bundle_id)
                        .contributed_components();
                    (contributed.as_ptr(), contributed.len())
                };
                // SAFETY: this is sound because the contributed_components Vec for explicit_bundle_id will not be accessed mutably as
                // part of init_dynamic_info. No mutable references will be created and the allocation will remain valid.
                self.init_dynamic_info(components, std::slice::from_raw_parts(ptr, len))
            };
```

cart commented on crates/bevy_ecs/src/bundle.rs:
```diff
@@ -1346,11 +1349,39 @@ impl Bundles {
                 // - it was created in the same order as the components in T
                 unsafe { BundleInfo::new(core::any::type_name::<T>(), components, component_ids, id) };
             bundle_infos.push(bundle_info);
+
+            // Be sure to update component bundle with requires
+            contributed_bundle_ids.remove(&TypeId::of::<T>());
+
```
Is this really necessary? I don't see how `contributed_bundle_ids` could exist for `T` if it hasn't already been registered. Unless I'm missing something, this feels like "defensive coding" protecting against incorrect internal code. Remove ops are expensive enough that we should remove this if it isn't functionally required. 

```suggestion
```

rewin123 replied:
It's defense programming. Really it's a slowdown for no apparent reason

cart reviewed (commented):
I'm on board for this approach, but I have a couple of optimization suggestions

rewin123 commented:
> I'm on board for this approach, but I have a couple of optimization suggestions

Thanks for the review! All remarks were justified and fixed

HackerFoo commented:
I wrote logic to handle enabling and disabling components with dependencies which may be useful as a reference, based on how package systems typically work:
* When enabling something, all dependencies are automatically enabled.
* When disabling something, anything depending on it is also disabled, as well as any dependencies that are only needed to support it.
The logic is in [`dtask_requested`](https://github.com/HackerFoo/dtask/blob/master/src/dtask.c#L86-L101). It relies on statically computed sets of (transitive) dependencies and dependents. Basically, you need to track the set of explicitly enabled/added components, and then from that compute the true set of components, and diff that with the current state.
